### DISCUSSION TITLE
a hybrid model using YOLOv11 and MobileNetV3 with an attention fusion mechanism

### DISCUSSION THREAD

**Discussion by wayanastari** [ultralytics/ultralytics#22215] [Q&A]
Hi Ultralytics team,
I am working on a research project where I want to create a hybrid model using YOLOv11 and MobileNetV3 with an attention fusion mechanism.
The pipeline I imagine is:
1.	Train YOLOv11n on my dataset to get detection results and global image features from the backbone.
2.	Extract ROIs (bounding boxes) from YOLO detections.
3.	Pass each ROI through MobileNetV3 to obtain local features.
4.	Fuse the global features (from YOLOv11 backbone) and local features (from MobileNetV3) using an attention module.
5.	Use the fused features for classification or refinement.
I am still confused on how to properly extract backbone features from YOLOv11.
•	Is there an official or recommended way to access intermediate feature maps or global feature vectors from the YOLOv11 backbone?
•	Should I modify the YOLO model YAML, or is there a built-in hook/forward feature extractor in the Ultralytics framework?
•	After extracting YOLO features, what’s the best practice to combine them with external models like MobileNetV3?
Could you please guide me or share an example of how to integrate YOLOv11 with an external CNN model for feature fusion?
Thank you so much!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wayanastari, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will also assist you here soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For your hybrid YOLO11 + MobileNetV3 feature-fusion idea, it would help the team if you could share:
- Your ultralytics package version and environment details (OS, CUDA/CUDNN, PyTorch)
- A short code snippet showing how you’re currently attempting to access YOLO11 backbone features
- Any model YAML changes you tried (if applicable)
- A brief description of the expected vs. actual features you’re getting, and any errors/warnings
- A small sample input and the corresponding outputs you wish to fuse

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Great idea, and you can do this without forking: grab the FPN maps right before the YOLO11 `Detect` head via a forward_pre_hook, pool them for a “global” vector, crop ROIs for MobileNetV3 “local” vectors, then fuse.

Python (feature tap + simple ROI encoder)
```python
import torch
import torch.nn.functional as F
from ultralytics import YOLO
from ultralytics.nn.modules.head import Detect
from torchvision.models import mobilenet_v3_large, MobileNet_V3_Large_Weights

yolo = YOLO("yolo11n.pt")  # Ultralytics YOLO11
neck = {}

def grab_neck_feats(module, inputs):
    x = inputs[0]  # list of [P3, P4, P5, ...] tensors
    neck["maps"] = [t if t.requires_grad else t.detach() for t in x]

detect = next(m for m in yolo.model.modules() if isinstance(m, Detect))
h = detect.register_forward_pre_hook(grab_neck_feats)

results = yolo(["img.jpg"])  # run once to populate neck["maps"]
h.remove()

# Global image descriptor from top pyramid map (e.g., P5)
global_vec = F.adaptive_avg_pool2d(neck["maps"][-1], 1).flatten(1)  # [B, C]

# ROI encoder with MobileNetV3 features -> vector per ROI
mb = mobilenet_v3_large(weights=MobileNet_V3_Large_Weights.DEFAULT).features.eval()
roi_encoder = torch.nn.Sequential(mb, torch.nn.AdaptiveAvgPool2d(1), torch.nn.Flatten())

# For each detected box, crop your ROI to a tensor (CxHxW) and run:
# roi_vec = roi_encoder(roi_tensor)  # then fuse with global_vec (concat/MLP/attention)
```

If you prefer a YAML-first approach, tap specific backbone outputs with `Index` and add ops like `nn.AdaptiveAvgPool2d`/`nn.Flatten`; see the Index example and the module resolution rules in the Model YAML guide at the “Index module for feature selection” and “Module resolution system” sections: [Model YAML Configuration Guide](https://docs.ultralytics.com/guides/model-yaml-config/). For general YOLO11 usage and shapes, see [Ultralytics YOLO11](https://docs.ultralytics.com/models/yolo11/).

If you hit any snags integrating the hook or YAML taps, share a minimal snippet (inputs, shapes, and where you want to fuse) here and we’ll iterate in-thread.

**Reply by wayanastari** [reply]
Hi! 
I’d like to clarify something about model usage:

After integrating both models (YOLO11 and MobileNetV3) as in your example, will the final hybrid system still require loading both models separately each time for inference, or is it possible to merge them into a single model file (e.g., one .pt that can be called directly)?

In other words, can this hybrid be trained and exported as one unified model, or is it intended to remain as two collaborating modules at runtime?

Thanks a lot for the guidance.  I just want to be sure about the best way to deploy this kind of fused model.

**Reply by glenn-jocher** [reply]
Short answer: both are possible—simplest is to keep YOLO11 and MobileNetV3 as two modules loaded at runtime; if you want a single artifact, wrap them in one `nn.Module` (or a custom YOLO head) that does `ROIAlign` + fusion inside `forward()`, then train end‑to‑end and export as a unified graph where the target backend supports your ops; see the supported targets and caveats in the [Export docs](https://docs.ultralytics.com/modes/export/), and note that `RoIAlign` exports cleanly to ONNX/OpenVINO while backends like TensorRT/CoreML/MNN may need plugins or adjustments (for MNN specifics, see the [MNN export guide](https://docs.ultralytics.com/integrations/mnn/)).
